File: 2024/bin/day09.py
import fileinput

lines = [[*map(int,list(line.strip()))] for line in fileinput.input() if line.strip()]
data = lines[0]

assert len(data) % 2 == 1

def calculate_checksum(memory):
    checksum = 0
    for i in range(len(memory)):
        checksum += memory[i] * i
    return checksum


def task1(data):
    data = data.copy()
    idx = 0
    file_idx = len(data) - 1
    memory = []

    while idx <= file_idx:
        if idx % 2 == 0:
            file_id = idx // 2
            for _ in range(data[idx]):
                memory.append(file_id)
            idx += 1
        elif idx % 2 == 1 and data[idx] == 0:
            idx += 1
        elif idx % 2 == 1 and data[file_idx] == 0:
            file_idx -= 2
        elif idx % 2 == 1 and data[idx] > 0:
            file_id = file_idx // 2
            movable_count = min(data[idx], data[file_idx])
            for _ in range(movable_count):
                memory.append(file_id)
            data[idx] -= movable_count
            data[file_idx] -= movable_count
    return calculate_checksum(memory)

print('Task 1:', task1(data))


def task2(data):
    data = data.copy()
    detailed_memory = []
    for i in range(len(data)):
        file_id = i // 2
        # (free_space: int , blocks [(file_id, len)])
        if i % 2 == 0:
            detailed_memory.append((0, [(file_id, data[i])]))
        else:
            detailed_memory.append((data[i], []))

    movable_file_idx = len(data) - 1
    while movable_file_idx > 1:
        if data[movable_file_idx] > 0:
            file_to_move = detailed_memory[movable_file_idx][1][-1]
            potential_free_space_idx = 1
            # find block block with enough free space
            while potential_free_space_idx < movable_file_idx:
                if detailed_memory[potential_free_space_idx][0] >= data[movable_file_idx]:
                    break
                potential_free_space_idx += 2

            # if found - move file there
            if potential_free_space_idx < movable_file_idx:
                detailed_memory[potential_free_space_idx] = (
                    detailed_memory[potential_free_space_idx][0] - file_to_move[1],
                    detailed_memory[potential_free_space_idx][1] + [file_to_move]
                )
                detailed_memory[movable_file_idx] = (
                    detailed_memory[movable_file_idx][0] + file_to_move[1],
                    detailed_memory[movable_file_idx][1][:-1],
                )
                pass
        movable_file_idx -= 2

    # restore memory from detailed memory layour
    memory = []
    for (free_space), blocks in detailed_memory:
        for (file_id, size) in blocks:
            memory.extend([file_id] * size)
        if free_space:
            memory.extend([0] * free_space)

    return calculate_checksum(memory)

print('Task 2:', task2(data))

File: 2024/bin/day09.ml
open Core
open Common

let data =
  read_lines_from_stdin ()
  |> List.hd_exn
  |> String.to_array
  |> Array.map ~f:(fun c -> Char.to_string c |> Int.of_string)
;;

assert (Array.length data mod 2 = 1)

let task1 data =
  let data = Array.copy data in
  let idx = ref 0 in
  let file_idx = ref (Array.length data - 1) in
  (* Precompute the total number of memory blocks to allocate *)
  let total_memory = Array.fold data ~init:0 ~f:( + ) in
  let memory = Array.create ~len:total_memory 0 in
  let mem_idx = ref 0 in
  while !idx <= !file_idx do
    if !idx mod 2 = 0
    then (
      (* Write blocks for file_id from idx *)
      let file_id = !idx / 2 in
      for _ = 1 to data.(!idx) do
        memory.(!mem_idx) <- file_id;
        incr mem_idx
      done;
      incr idx)
    else if !idx mod 2 = 1 && data.(!idx) = 0
    then incr idx
    else if !idx mod 2 = 1 && data.(!file_idx) = 0
    then file_idx := !file_idx - 2
    else if !idx mod 2 = 1 && data.(!idx) > 0
    then (
      let file_id = !file_idx / 2 in
      let movable_count = Int.min data.(!idx) data.(!file_idx) in
      for _ = 1 to movable_count do
        memory.(!mem_idx) <- file_id;
        incr mem_idx
      done;
      data.(!idx) <- data.(!idx) - movable_count;
      data.(!file_idx) <- data.(!file_idx) - movable_count)
  done;
  (* Calculate the checksum *)
  Array.foldi memory ~f:(fun i acc file_id -> acc + (file_id * i)) ~init:0
;;

Stdio.printf "Task 1: %d\n" (task1 data);;
Out_channel.flush Out_channel.stdout

(* Task 2 is implemented only in Python (day09.py) because
   OCaml version was quite slow due to lots of memory allocations. (immutable lists)
   And the code is very messy if I try to optimize it by using 2d arrays.:shrug: *)
